src/java.net.http/share/classes/jdk/internal/net/http/common/HttpBodySubscriberWrapper.java

```diff
@@ -61,12 +62,33 @@ public void cancel() { }
     final BodySubscriber<T> userSubscriber;
     final AtomicBoolean completed = new AtomicBoolean();
     final AtomicBoolean subscribed = new AtomicBoolean();
-    volatile Subscription subscription;
+    volatile SubscriptionWrapper subscription;
     volatile Throwable withError;
     public HttpBodySubscriberWrapper(BodySubscriber<T> userSubscriber) {
         this.userSubscriber = userSubscriber;
     }
 
+    private class SubscriptionWrapper implements Subscription {
+        final Subscription subscription;
+        SubscriptionWrapper(Subscription s) {
+            this.subscription = Objects.requireNonNull(s);
+        }
+        @Override
+        public void request(long n) {
+            subscription.request(n);
+        }
+
+        @Override
+        public void cancel() {
+            try {
+                subscription.cancel();
+                onCancel();
+            } catch (Throwable t) {
+                onError(t);
+            }
```

Comment:
Thank you for making this change. Do you think, we should only call `onError()` if `subscription.cancel()` fails? In its current form above, it will end up calling the `onError()` if the `onCancel()` fails for whatever reason.